rfcs/2020-03-06-1999-api-extensions-for-lua-transform.md

```diff
@@ -0,0 +1,204 @@
+        kind = "absolute",
+        tags = {
+          host = "localhost"
+        },
+        value = {
+          type = "counter",
+          value = 24.2
+        }
+      }
+    }
+  """
+  [[transforms.lua.timers]]
+  interval = 60
+  source = """
+    event = {
+      log = {
+        message = "heartbeat",
+        timestamp = os.date("!*t),
+      }
+    }
+    emit(event, "auxiliary")
+  """
+```
+
+The code above consumes the incoming events, counts them, and then emits these stats about these counts every 10 seconds. In addition, it sends debug logs about its functioning into a separate lane called `auxiliary`.
+
+### Proposed changes
+
+* Add `version` configuration option which would allow the users to chose between the new API described in this RFC (version 2) and the old one (version 1).
+* Hooks for initialization and shutdown called `init` and `shutdown`. They are defined as strings of Lua code in the `hooks` section of the configuration of the transform.
+* Timers which define pieces of code that are executed periodically. They are defined in array `timers`, each timer takes two configuration options: `interval` which is the interval for execution in seconds and `source` which is the code which is to be executed periodically.
+* Events are produced by the transform by calling function `emit` with the first argument being the event and the second option argument being the name of the lane where to emit the event. Outputting the events by storing them to the `event` global variable should not be supported, so its content would be ignored.
+* Support direct access to the nested fields (in both maps and arrays).
+* Add support for the timestamp type as a `userdata` object with the same visible fields as in the table returned by [`os.date`](https://www.lua.org/manual/5.3/manual.html#pdf-os.date). In addition, monkey-patch `os.date` function available inside Lua scripts to make it return the same kind of userdata instead of a table if it is called with `*t` or `!*t` as the argument. This is necessary to allow one-to-one correspondence between types in Vector and Lua.
+
+## Sales Pitch
+
+The proposal
+
+* gives users more power to create custom transforms;
+* supports both logs and metrics;
+* makes it possible to add complexity to the configuration of the transform gradually only when needed.
+
+## Drawbacks
+
+The only drawback is that supporting both dot notation and classical indexing makes it impossible to add escaping of dots in field names. For example, for incoming event structure like
+
+```json
+{
+  "field.first": {
+    "second": "value"
+  }
+}
+```
+
+accessing `event["field.first"]` would return `nil`.
+
+However, because of the specificity of the observability data, there seems to be no need to have both field names with dots and nested fields.
+
+## Outstanding Questions
+
+* Should timestamps be automatically inserted to created logs and metrics created as tables inside the transform is they are not present?
```

Comment:
I don't see why they should...

Reply:
For example, `stdin` source inserts timestamps to newly created events automatically.

Reply:
It should not be possible for a source to emit an event without a `timestamp` field, so the `lua` transform should not concern itself with this. If a user explicitly removes the `timestamp` field then they'll deal with problems, but we should not cater to that edgecase.

Reply:
The question is about newly created events inside the transforms, as the user can forget to add the timestamp to them.

Reply:
Ah. Yes, in that case, we should populate the appropriate [schema fields](https://vector.dev/docs/about/data-model/log/#schema). The names should respect those provided as part of [global `log_schema` setting](https://vector.dev/docs/reference/global-options/#log_schema). I'd argue that both `timestamp` and `host` should be automatically set.